en/config_mc/filter_tuning.md

```diff
@@ -0,0 +1,98 @@
+# MC Filter Tuning & Control Latency
+
+Filters can be used to trade off [control latency](#control-latency), which affects flight performance, and noise filtering, which impacts both flight performance and motor health.
+
+This topic provides an overview of control latency and PX4 filter tuning.
+
+:::note
```

Comment:
@bkueng So re your comment in https://github.com/PX4/PX4-user_guide/pull/1181#pullrequestreview-635322101

1. As this is more or less MC specific I added this note to repeat the "do basic PID tuning but undertune" message. I also moved the whole doc from advanced config to MC config. 

The notch filters sounds interesting. Where do these fit in the pipeline - ie are they "use this instead of IMU_GYRO_CUTOFF" or might you use all of them?
My very old memory is that a notch filter allows you to filter out a specified band of frequencies, so I guess you are saying that if you do a frequency analysis and see that all your noise is at one frequency you use a notch filter to get rid of that. If it is spread then you might just use the normal IMU_GYRO_CUTOFF.

Anyway, what I think we need here is for **you** to add the information about tuning for the new parameters, including an FFT, list of the parameters affected. Otherwise this will just be a second class citizen in the doc. Can ou do that?

With respect to the dynamic notch filter it sounds premature to fully document it. So perhaps we should just have a note that it exists but you should not yet use it - i.e. don't touch parameters X, Y, Z. 

Make sense?

My thinking is that we would merge this, and you could add update for notch "in the coming month"

Reply:
You're correct. A notch can be used in addition: you would set the general low-pass quite high (e.g. 120), and then add a lower notch at the rotor blade pass frequency, or your peak, e.g. 80.
Params are IMU_GYRO_NF_BW and IMU_GYRO_NF_FREQ.

This is an example:
![bokeh_plot](https://user-images.githubusercontent.com/281593/114991127-8bf1da00-9e99-11eb-9c6a-bb3059f36792.png)
You would set the freq at 32, and bandwidth to 5 (it's often not quite that narrow).

Does that give you enough info?

> With respect to the dynamic notch filter it sounds premature to fully document it

I think so too.

Reply:
> Does that give you enough info?

@bkueng Nearly. I've taken a first shot at this, but I expect it to need work.

The open questions are:
1.  **when** you do the notch filter 
    - my guess is that you always start by reviewing the plot from with default filter settings. If you saw a spike like this one the FIRST thing you would do is set the notch filter. I.e you wouldn't do the other tuning until any notch was removed right?

2. if you use a notch filter how you determine the Low pass cutoff. 
   - my guess is that for determining the low pass cuttoff you use the **same** process that you used if there was no notch (see next question). The point of the notch is to put you in a state where you don't have to put the low pass filter before that noise spike - after you have it the signal is "as though you never had a spike".

3.  irrespective of whether you use a notch filter or not, at what cuttoff do you start with for IMU_GYRO_CUTOFF (for the d term you said 30).
     - you show plots for IMU_GYRO_CUTOFF of 40, 70, 90, where the mean for the plot is 40, and the signal appears to trail off into noise at about 70. The analysis showed that above 70 noise increases.
     - So my assumption might be that a good stating point for setting the IMU_GYRO_CUTOFF  will be somewhere between the mean and where you see your main frequencies trailing off. Is there a "rule of thumb?

> A notch can be used in addition: you would set the general low-pass quite high (e.g. 120), and then add a lower notch at the rotor blade pass frequency, or your peak, e.g. 80.

Are you saying that the general LP cuttoff might be set at either at 120 or 80? Or are you suggesting IMU_GYRO_CUTOFF =120 and D term cutoff=80? 

Hope that all makes sense.

Reply:
Hi @hamishwillee , in case it is useful for future questions as well as others in a general understanding, these notch filters pretty much mathematically follow what is called (as perhaps a misnomer), "RPM filtering" in betaflight: https://oscarliang.com/rpm-filter/

Reply:
In a nutshell, (no PX4 pun, but it is fun anyway...), each airframe will have harmonic resonances which can be seen as yellow bands in the px4 flight review Acceleration PSD graph.  Among other resonance, filtering these out helps reject disturbances that the controller will act upon.  You probably already were aware of this and just clarifying, but to add gravity to what you are documenting I wanted to add a comment.

Reply:
> my guess is that you always start by reviewing the plot from with default filter settings

The defaults are set really low, so you'd increase it at first to see 'how bad it is'.

> I.e you wouldn't do the other tuning until any notch was removed right?

Right

> my guess is that for determining the low pass cuttoff you use the same process that you used if there was no notch (see next question). The point of the notch is to put you in a state where you don't have to put the low pass filter before that noise spike - after you have it the signal is "as though you never had a spike".

Exactly. You set it the next higher peak/where it generally goes up (there could be multiple notch filters, but there's currently only one to configure, as often there is one lower peak).

> So my assumption might be that a good stating point for setting the IMU_GYRO_CUTOFF will be somewhere between the mean and where you see your main frequencies trailing off. Is there a "rule of thumb?

Personally I just start quite high (around 100) and see how it goes (with some experience you can hear it, but also check if the motors get hot), then go either further up or down. You can do the same based on plots.

> Are you saying that the general LP cuttoff might be set at either at 120 or 80? Or are you suggesting IMU_GYRO_CUTOFF =120 and D term cutoff=80?

Here I meant IMU_GYRO_CUTOFF =120 and IMU_GYRO_NF_FREQ = 80, ignoring the D filter for that example.
Might be worth mentioning that the d filter should not be too far off from IMU_GYRO_CUTOFF, as an increased difference in delay starts to have negative impacts at some point. You need to go quite far for that though, e.g. d=15, gyro=80.

> In a nutshell, (no PX4 pun, but it is fun anyway...)

:D

Reply:
> Here I meant IMU_GYRO_CUTOFF =120 and IMU_GYRO_NF_FREQ = 80, ignoring the D filter for that example.

@bkueng That makes no sense. I can see why you might set the IMU_GYRO_CUTOFF =120 (because the noise is still low) but the spike is at 32, so why would you set `IMU_GYRO_NF_FREQ = 80`? Or were you talking about some other hypothetical example where the spike is at 80?

![image](https://user-images.githubusercontent.com/5368500/115640470-6281f400-a35a-11eb-9626-039686aa1bda.png)

Reply:
@bkueng Just to be absolutely clear, are the tuning of the notch and IMU_GYRO_CUTOFF from the logs independent? 

I.e. if I do the first flights to tune IMU_GYRO_CUTOFF without tuning the notch I will end up with a particular value for cuttoff based at the point where the noise after 60Hz starts to increase. If I were to run the test once, see I have a low frequency spike and set the notch, THEN run the flights to tune IMU_GYRO_CUTOFF will I get a different value for IMU_GYRO_CUTOFF?

If they are independent then the instructions as they are now are OK. However if you have to do the notch filter first (as I imply from your previous answers) then I will need to reorder the instructions.

----

Further, you describe an experienced approach, which is "set the IMU_GYRO_CUTOFF high (e.g. 100Hz) to assess if there is a problem, which you can hear or detect from the motors getting hot. Then increase/decrease values until you no longer hear the whine or feel excessive heat". Presumably if you set at 1kHz and the motors are still cool/no odd sound then no tuning is needed :-) 
- About right?
- Is there a value for the cuttoff at which you might say "good enough/no point tuning for latency". In other words, if you ran a test at 100Hz and you saw that the motors were cool and there was no bad sound, would you just stop? Or do you always tune until there is a problem?
- How can you tell from this if you need a notch filter?

The existing process is pretty much dumb/methodical - you iterate up your cuttoff from default in 10Hz increments and look at the plots at the end. This is good because no expertise is required. Is there a benefit for these users in doing an initial high test and checking logs in this case? Or is the process pretty OK?

Essentially I'm trying to work out if there is any benefit for the broad base of users in starting from a higher test value for the cuttoff, and what rule of thumb would we use for that. The current process involves quite a few tests, but it is pretty straightforward, and you walk away from it with all the information needed to do your tuning.

Reply:
@mcsauder Thanks for commenting - that is certainly useful for me to know.

So here I refer to the low frequency noise spike which we see in the FFTs. The example given is due to rotor blade wash. I infer from your comment that  if I was to look at a corresponding power spectral density plot I might see a yellow band at this frequency, similar to those show in the [Flight Review analysis doc here](https://docs.px4.io/master/en/log/flight_review.html#acceleration-power-spectral-density). Is that about right?

If so, I was kind aware of this. Do you think perhaps we should add a note somewhere also pointing to the Flight review analysis doc? Or is there more you think I could/should say?

In particular, are there other sources of harmonics I might mention?

Reply:
> Or were you talking about some other hypothetical example where the spike is at 80?

Yes unrelated to the example plot.

> I.e. if I do the first flights to tune IMU_GYRO_CUTOFF without tuning the notch I will end up with a particular value for cuttoff based at the point where the noise after 60Hz starts to increase. If I were to run the test once, see I have a low frequency spike and set the notch, THEN run the flights to tune IMU_GYRO_CUTOFF will I get a different value for IMU_GYRO_CUTOFF?

They are independent.

> Presumably if you set at 1kHz and the motors are still cool/no odd sound then no tuning is needed :-)

Indeed, but I'm not aware of any vehicle requiring no filtering at all. Would be nice though, let me know if you find one :)

> Is there a value for the cuttoff at which you might say "good enough/no point tuning for latency". In other words, if you ran a test at 100Hz and you saw that the motors were cool and there was no bad sound, would you just stop? Or do you always tune until there is a problem?

It depends on the vehicle size and expectations. I.e. FPV racers go as high as they can, but for me personally, flying line-of-sight, ~120 is good enough (and never needed a notch), while having the D a bit lower (50-80).
For bigger vehicles you probably don't have to go beyond 80 (as latency is less critical).

> How can you tell from this if you need a notch filter?

For that you really do need to look at the plots.

> Is there a benefit for these users in doing an initial high test and checking logs in this case?

You can start high as well, just need to be aware of the risks. I'd say don't fly for more than 20-30sec. And if the vibrations getting through are really exceedingly high, there's a risk of crashing the vehicle as it becomes unflyable.
Eventually I hope we can simplify this with using dynamic (notch) filters.

Reply:
Thanks @bkueng 

Awesome. I've added the additional info mostly as a "tips" section. Can you please do a final sanity check. If it is OK I will merge tomorrow and we can deal with any further comments as a post process.